File: BOJ/115.Greedy/1931.py
# 회의실배정
import sys
input = sys.stdin.readline
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
arr.sort(key = lambda k : (k[1], k[0])) # 회의 끝나는 시간 오름차순 정렬 후, 시작 시간 오름차순정렬
def solution():
    ans = 0
    s = -1
    for i in range(n):
        if arr[i][0] >= s:
            s = arr[i][1]
            ans += 1
    print(ans)
solution()

# Test Case

# 5
# 4 4
# 4 4
# 3 4
# 2 4
# 1 4

# 답 : 3

# 5
# 0 1
# 1 2
# 2 2
# 2 3
# 3 3

# 답 : 5

# 5
# 4 5
# 4 4
# 3 4
# 2 5
# 1 3

# 답 : 4

# 3
# 2 3
# 3 5
# 4 4

# 답 : 2

File: BOJ/115.Greedy/1931.swift
// 회의실 배정
// 문제 유형: 그리디, 정렬
// 이전에 파이썬으로 풀었었는데 프로그래머스에서 이와 비슷한 문제를 접했음.
// 그런데 다시봐도 왜 이렇게 풀었는지 이해가 안 됐음..
// 정렬을 할 때 끝시간 오름 차순 후, 시작 시간 오름 차순 정렬을 해야 함.
// 그 이유는 끝시간을 기준으로 오름차순으로 정렬해야 이전 강의의 끝 시간을 기준으로 다음 강의의 시작 시간과 비교해서 끝시간을 갱신할 수 있음.
// 또한 시작 시간 오름 차순까지 정렬을 해야하는 이유는 강의 시작 시간이 큰 것부터 계산하게 되면 작은 것이 포함되지 않을 수 있기 때문
// 반례 예시
// 5
// 1 2
// 5 9
// 7 11
// 13 13
// 12 13
// 답은 4가 나와야하나 3이 나옴

import Foundation

let n = Int(readLine()!)!

print(solution())

func solution() -> Int {
    let array = input()
    var answer = 0
    var s = -1
    
    // 이전 강의의 끝 시간보다 다음 강의 시작 시간이 더 클 때만 그 시간의 끝시간으로 갱신 후, answer += 1
    for time in array {
        if time[0] >= s {
            s = time[1]
            answer += 1
        }
    }
    return answer
}

func input() -> [[Int]] {
    var newArray = [[Int]]()
    for _ in 0..<n {
        let split = readLine()!.components(separatedBy: " ").map { Int($0)! }
        newArray.append([split[0], split[1]])
    }
    // 끝시간 오름차순으로 정렬 후, 시작 시간 오름차순 정렬
    newArray.sort(by: { ($0[1], $0[0]) < ($1[1], $1[0]) })
    return newArray
}
